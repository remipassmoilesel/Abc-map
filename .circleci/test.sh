#!/usr/bin/env bash

set -x
set -e

# TODO: add tests

mkdir -p /data/db && /usr/bin/mongod &

cd shared && npm install && npm run build

cd ..

cd api && npm install && npm run build

cd ..

cd gui && npm install && npm run build

