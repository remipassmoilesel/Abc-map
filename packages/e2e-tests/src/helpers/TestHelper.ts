/**
 * Copyright © 2023 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import Chainable = Cypress.Chainable;
import { Routes } from './Routes';

export class TestHelper {
  public static init(): Chainable<any> {
    return (
      cy
        // We set viewport
        .viewport(1980, 1080)
        // We set default language
        .visit(Routes.map().format(), {
          onBeforeLoad(win) {
            Object.defineProperty(win.navigator, 'language', { value: 'en-US' });
            Object.defineProperty(win.navigator, 'languages', { value: ['en'] });
            Object.defineProperty(win.navigator, 'accept_languages', { value: ['en'] });
          },
        })
        // We clear local storage data (and redux store persistence)
        // Despite docs says it run automatically before each test, it does not clear side menu states
        .clearLocalStorage()
    );
  }
}
