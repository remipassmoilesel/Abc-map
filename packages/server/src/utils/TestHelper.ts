/**
 * Copyright © 2023 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import {
  AbcArtefact,
  AbcProjectManifest,
  AbcUser,
  ArtefactManifest,
  ArtefactType,
  CompressedProject,
  DEFAULT_PROJECTION,
  Language,
  LayerType,
  PredefinedLayerModel,
  ProjectConstants,
  Zipper,
} from '@abc-map/shared';
import * as uuid from 'uuid-random';
import { ProjectDocument } from '../projects/ProjectDocument';
import { DateTime } from 'luxon';
import { ArtefactManifestWithPath } from '../data-store/ArtefactManifestSchema';

export class TestHelper {
  public static sampleUser(): AbcUser {
    return {
      id: uuid(),
      email: `user-${uuid()}@test.ts`,
      password: 'what is a wr0ng passW0rd ????',
    };
  }

  public static sampleProject(): AbcProjectManifest {
    return {
      metadata: {
        id: uuid(),
        version: ProjectConstants.CurrentVersion,
        name: `Test project ${uuid()}`,
        public: false,
      },
      layers: [
        {
          type: LayerType.Predefined,
          metadata: {
            id: uuid(),
            name: 'OpenStreetMap',
            type: LayerType.Predefined,
            visible: true,
            active: false,
            opacity: 1,
            model: PredefinedLayerModel.OSM,
          },
        },
        {
          type: LayerType.Vector,
          metadata: {
            id: uuid(),
            name: 'Vecteurs',
            type: LayerType.Vector,
            visible: true,
            active: true,
            opacity: 1,
          },
          features: {
            type: 'FeatureCollection',
            features: [
              {
                id: uuid(),
                bbox: [1, 2, 3, 4],
                type: 'Feature',
                properties: {
                  val: 'var',
                },
                geometry: {
                  type: 'Point',
                  coordinates: [1, 5],
                },
              },
            ],
          },
        },
      ],
      layouts: {
        abcMapAttributionsEnabled: true,
        list: [],
      },
      view: {
        center: [1, 2],
        projection: DEFAULT_PROJECTION,
        resolution: 1000,
        rotation: 0,
      },
      sharedViews: {
        list: [],
        fullscreen: false,
        mapDimensions: {
          width: 800,
          height: 600,
        },
      },
    };
  }

  public static async sampleCompressedProject(): Promise<CompressedProject<Buffer>> {
    const project = this.sampleProject();
    const metadata = project.metadata;
    const zip = await Zipper.forNodeJS().zipFiles([{ path: ProjectConstants.ManifestName, content: Buffer.from(JSON.stringify(project), 'utf-8') }]);
    return {
      metadata: metadata,
      project: zip,
    };
  }

  public static sampleProjectDocument(): ProjectDocument {
    return {
      _id: uuid(),
      ownerId: uuid(),
      name: 'Fake project',
      version: '0.1',
      public: false,
    };
  }

  public static sampleArtefact(): AbcArtefact {
    return {
      id: uuid(),
      name: [{ language: Language.French, text: 'Burgers around the world' }],
      type: ArtefactType.Vector,
      license: 'AGPLv3.txt',
      attributions: ['Copyright somebody somewhere inc'],
      description: [{ language: Language.English, text: 'A beautiful artefact' }],
      files: ['/datastore/artefact/file1.gpx', '/datastore/artefact/file2.shp'],
      path: '/datastore/artefact/manifest.yml',
      provider: 'Nowhere inc',
      link: 'http://nowhere.net',
      previews: ['preview1.png'],
      keywords: [{ language: Language.English, text: ['beautiful', 'artifact'] }],
    };
  }

  public static sampleArtefactManifestWithPath(name: string): ArtefactManifestWithPath {
    return {
      version: '0.0.1',
      path: `/datastore/${name}/manifest.yml`,
      artefact: {
        name: [{ language: Language.English, text: name }],
        type: ArtefactType.BaseMap,
        license: `${name}-AGPLv3.txt`,
        attributions: ['Copyright somebody somewhere inc'],
        description: [{ language: Language.English, text: `A beautiful artefact named ${name}` }],
        keywords: [{ language: Language.English, text: ['beautiful', 'artifact', name] }],
        files: [`file1.gpx`, `file2.shp`],
        provider: `Nowhere inc`,
        link: `http://nowhere.net/${name}`,
      },
    };
  }

  public static sampleArtefactManifest(name: string): ArtefactManifest {
    return {
      version: '0.0.1',
      artefact: {
        name: [{ language: Language.English, text: name }],
        type: ArtefactType.BaseMap,
        license: `${name}-AGPLv3.txt`,
        attributions: ['Copyright somebody somewhere inc'],
        description: [{ language: Language.English, text: `A beautiful artefact named ${name}` }],
        keywords: [{ language: Language.English, text: ['beautiful', 'artifact', name] }],
        files: [`file1.gpx`, `file2.shp`],
        provider: `Nowhere inc`,
        link: `http://nowhere.net/${name}`,
      },
    };
  }

  public static randomDate(start: Date = new Date(1980, 1, 1, 0, 0, 0), end: Date = new Date(2020, 1, 1, 0, 0, 0)): DateTime {
    const timestamp = start.getTime() + Math.random() * (end.getTime() - start.getTime());
    return DateTime.fromMillis(timestamp);
  }

  /**
   * Warning: use this is a very bad idea
   */
  public static wait(timeMs: number): Promise<void> {
    return new Promise((resolve) => {
      setTimeout(resolve, timeMs);
    });
  }
}
