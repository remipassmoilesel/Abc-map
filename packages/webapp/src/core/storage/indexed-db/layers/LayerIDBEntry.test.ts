/**
 * Copyright © 2023 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import { CURRENT_VERSION, LayerIDBEntry } from './LayerIDBEntry';
import { TestHelper } from '../../../utils/test/TestHelper';

describe('LayerIDBEntry', () => {
  it('should not change without migration', () => {
    const entry1: LayerIDBEntry = {
      version: CURRENT_VERSION,
      layer: {
        ...TestHelper.sampleOsmLayer(),
        metadata: {
          ...TestHelper.sampleOsmLayer().metadata,
          id: 'test-layer-1',
        },
      },
    };

    expect(entry1).toMatchSnapshot();

    const entry2: LayerIDBEntry = {
      version: CURRENT_VERSION,
      layer: {
        ...TestHelper.sampleWmsLayer(),
        metadata: {
          ...TestHelper.sampleWmsLayer().metadata,
          id: 'test-layer-1',
        },
      },
    };

    expect(entry2).toMatchSnapshot();

    const entry3: LayerIDBEntry = {
      version: CURRENT_VERSION,
      layer: {
        ...TestHelper.sampleWmtsLayer(),
        metadata: {
          ...TestHelper.sampleWmtsLayer().metadata,
          id: 'test-layer-1',
        },
      },
    };

    expect(entry3).toMatchSnapshot();

    const entry4: LayerIDBEntry = {
      version: CURRENT_VERSION,
      layer: {
        ...TestHelper.sampleXyzLayer(),
        metadata: {
          ...TestHelper.sampleXyzLayer().metadata,
          id: 'test-layer-1',
        },
      },
    };

    expect(entry4).toMatchSnapshot();

    const entry5: LayerIDBEntry = {
      version: CURRENT_VERSION,
      layer: {
        ...TestHelper.sampleVectorLayer(),
        metadata: {
          ...TestHelper.sampleVectorLayer().metadata,
          id: 'test-layer-1',
        },
        features: {
          ...TestHelper.sampleVectorLayer().features,
          features: TestHelper.sampleVectorLayer().features.features.map((feat, i) => ({ ...feat, id: `test-feature-${i}` })),
        },
      },
    };

    expect(entry5).toMatchSnapshot();
  });
});
