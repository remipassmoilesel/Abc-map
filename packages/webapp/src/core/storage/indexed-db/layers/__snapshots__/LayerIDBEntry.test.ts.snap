// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LayerIDBEntry should not change without migration 1`] = `
Object {
  "layer": Object {
    "metadata": Object {
      "active": false,
      "id": "test-layer-1",
      "model": "OSM",
      "name": "OpenStreetMap",
      "opacity": 1,
      "type": "Predefined",
      "visible": true,
    },
    "type": "Predefined",
  },
  "version": 2,
}
`;

exports[`LayerIDBEntry should not change without migration 2`] = `
Object {
  "layer": Object {
    "metadata": Object {
      "active": true,
      "auth": Object {
        "password": "test-password",
        "username": "test-username",
      },
      "extent": Array [
        1,
        2,
        3,
        4,
      ],
      "id": "test-layer-1",
      "name": "Couche WMS",
      "opacity": 1,
      "projection": Object {
        "name": "EPSG:4326",
      },
      "remoteLayerName": "test-layer-name",
      "remoteUrls": Array [
        "http://domain.fr/wms",
      ],
      "type": "Wms",
      "visible": true,
    },
    "type": "Wms",
  },
  "version": 2,
}
`;

exports[`LayerIDBEntry should not change without migration 3`] = `
Object {
  "layer": Object {
    "metadata": Object {
      "active": true,
      "auth": Object {
        "password": "test-password",
        "username": "test-username",
      },
      "capabilitiesUrl": "http://domain.fr/wmts",
      "id": "test-layer-1",
      "name": "Couche WMTS",
      "opacity": 1,
      "remoteLayerName": "GEOGRAPHICALGRIDSYSTEMS.MAPS",
      "sourceOptions": Object {
        "crossOrigin": undefined,
        "dimensions": Object {},
        "format": "image/jpeg",
        "layer": "GEOGRAPHICALGRIDSYSTEMS.MAPS",
        "matrixSet": "PM",
        "projection": EPSG3857Projection {
          "axisOrientation_": "enu",
          "canWrapX_": true,
          "code_": "EPSG:3857",
          "defaultTileGrid_": null,
          "extent_": Array [
            -20037508.342789244,
            -20037508.342789244,
            20037508.342789244,
            20037508.342789244,
          ],
          "getPointResolutionFunc_": [Function],
          "global_": true,
          "metersPerUnit_": undefined,
          "units_": "m",
          "worldExtent_": Array [
            -180,
            -85,
            180,
            85,
          ],
        },
        "requestEncoding": "KVP",
        "style": "normal",
        "tileGrid": WMTSTileGrid {
          "extent_": Array [
            -7034346.839028204,
            -2443080.2980507985,
            6216929.25854652,
            6652314.571658893,
          ],
          "fullTileRanges_": Array [
            TileRange {
              "maxX": 0,
              "maxY": 0,
              "minX": 0,
              "minY": 0,
            },
            TileRange {
              "maxX": 1,
              "maxY": 1,
              "minX": 0,
              "minY": 0,
            },
            TileRange {
              "maxX": 2,
              "maxY": 2,
              "minX": 1,
              "minY": 1,
            },
            TileRange {
              "maxX": 5,
              "maxY": 4,
              "minX": 2,
              "minY": 2,
            },
            TileRange {
              "maxX": 10,
              "maxY": 8,
              "minX": 5,
              "minY": 5,
            },
            TileRange {
              "maxX": 20,
              "maxY": 17,
              "minX": 10,
              "minY": 10,
            },
            TileRange {
              "maxX": 41,
              "maxY": 35,
              "minX": 20,
              "minY": 21,
            },
            TileRange {
              "maxX": 83,
              "maxY": 71,
              "minX": 41,
              "minY": 42,
            },
            TileRange {
              "maxX": 167,
              "maxY": 143,
              "minX": 83,
              "minY": 85,
            },
            TileRange {
              "maxX": 335,
              "maxY": 287,
              "minX": 166,
              "minY": 171,
            },
            TileRange {
              "maxX": 670,
              "maxY": 574,
              "minX": 332,
              "minY": 342,
            },
            TileRange {
              "maxX": 1341,
              "maxY": 1148,
              "minX": 664,
              "minY": 684,
            },
            TileRange {
              "maxX": 2683,
              "maxY": 2297,
              "minX": 1329,
              "minY": 1368,
            },
            TileRange {
              "maxX": 5366,
              "maxY": 4595,
              "minX": 2658,
              "minY": 2736,
            },
            TileRange {
              "maxX": 10733,
              "maxY": 9190,
              "minX": 5316,
              "minY": 5472,
            },
            TileRange {
              "maxX": 21467,
              "maxY": 18381,
              "minX": 10632,
              "minY": 10944,
            },
            TileRange {
              "maxX": 42934,
              "maxY": 36763,
              "minX": 21264,
              "minY": 21889,
            },
            TileRange {
              "maxX": 85869,
              "maxY": 73526,
              "minX": 42529,
              "minY": 43778,
            },
            TileRange {
              "maxX": 171738,
              "maxY": 147052,
              "minX": 85058,
              "minY": 87557,
            },
          ],
          "matrixIds_": Array [
            "0",
            "1",
            "2",
            "3",
            "4",
            "5",
            "6",
            "7",
            "8",
            "9",
            "10",
            "11",
            "12",
            "13",
            "14",
            "15",
            "16",
            "17",
            "18",
          ],
          "maxZoom": 18,
          "minZoom": 0,
          "origin_": null,
          "origins_": Array [
            Array [
              -20037508.342789248,
              20037508.342789248,
            ],
            Array [
              -20037508.342789248,
              20037508.342789248,
            ],
            Array [
              -20037508.342789248,
              20037508.342789248,
            ],
            Array [
              -20037508.342789248,
              20037508.342789248,
            ],
            Array [
              -20037508.342789248,
              20037508.342789248,
            ],
            Array [
              -20037508.342789248,
              20037508.342789248,
            ],
            Array [
              -20037508.342789248,
              20037508.342789248,
            ],
            Array [
              -20037508.342789248,
              20037508.342789248,
            ],
            Array [
              -20037508.342789248,
              20037508.342789248,
            ],
            Array [
              -20037508.342789248,
              20037508.342789248,
            ],
            Array [
              -20037508.342789248,
              20037508.342789248,
            ],
            Array [
              -20037508.342789248,
              20037508.342789248,
            ],
            Array [
              -20037508.342789248,
              20037508.342789248,
            ],
            Array [
              -20037508.342789248,
              20037508.342789248,
            ],
            Array [
              -20037508.342789248,
              20037508.342789248,
            ],
            Array [
              -20037508.342789248,
              20037508.342789248,
            ],
            Array [
              -20037508.342789248,
              20037508.342789248,
            ],
            Array [
              -20037508.342789248,
              20037508.342789248,
            ],
            Array [
              -20037508.342789248,
              20037508.342789248,
            ],
          ],
          "resolutions_": Array [
            156543.033928041,
            78271.51696402048,
            39135.758482010235,
            19567.87924100512,
            9783.93962050256,
            4891.96981025128,
            2445.98490512564,
            1222.99245256282,
            611.49622628141,
            305.7481131407048,
            152.8740565703525,
            76.43702828517624,
            38.21851414258813,
            19.10925707129406,
            9.554628535647032,
            4.777314267823516,
            2.388657133911758,
            1.194328566955879,
            0.5971642834779395,
          ],
          "tileSize_": null,
          "tileSizes_": Array [
            256,
            256,
            256,
            256,
            256,
            256,
            256,
            256,
            256,
            256,
            256,
            256,
            256,
            256,
            256,
            256,
            256,
            256,
            256,
          ],
          "tmpExtent_": Array [
            0,
            0,
            0,
            0,
          ],
          "tmpSize_": Array [
            256,
            256,
          ],
          "zoomFactor_": undefined,
        },
        "urls": Array [
          "https://wxs.ign.fr/0666d69045a3e461-lol/geoportail/wmts?",
        ],
        "wrapX": false,
      },
      "type": "Wmts",
      "visible": true,
    },
    "type": "Wmts",
  },
  "version": 2,
}
`;

exports[`LayerIDBEntry should not change without migration 4`] = `
Object {
  "layer": Object {
    "metadata": Object {
      "active": true,
      "id": "test-layer-1",
      "name": "Couche XYZ",
      "opacity": 1,
      "projection": Object {
        "name": "EPSG:4326",
      },
      "remoteUrl": "http://domain.fr/xyz/{x}/{y}/{z}",
      "type": "Xyz",
      "visible": true,
    },
    "type": "Xyz",
  },
  "version": 2,
}
`;

exports[`LayerIDBEntry should not change without migration 5`] = `
Object {
  "layer": Object {
    "features": Object {
      "features": Array [
        Object {
          "bbox": Array [
            1,
            2,
            3,
            4,
          ],
          "geometry": Object {
            "coordinates": Array [
              1,
              5,
            ],
            "type": "Point",
          },
          "id": "test-feature-0",
          "properties": Object {
            "val": "var",
          },
          "type": "Feature",
        },
      ],
      "type": "FeatureCollection",
    },
    "metadata": Object {
      "active": true,
      "id": "test-layer-1",
      "name": "Vecteurs",
      "opacity": 1,
      "type": "Vector",
      "visible": true,
    },
    "type": "Vector",
  },
  "version": 2,
}
`;
